cmd/clair/introspection.go

```diff
@@ -0,0 +1,184 @@
+	mux := http.NewServeMux()
+	srv := http.Server{
+		Addr: cfg.IntrospectionAddr,
+	}
+	log := zerolog.Ctx(ctx).With().Logger()
+
+	// Metrics config
+	if n := cfg.Metrics.Name; n != "" {
+		log.Info().Str("sink", n).Msg("configuring metrics sink")
+	}
+	switch cfg.Metrics.Name {
+	case "":
+		log.Info().Str("sink", "default").Msg("configuring metrics sink")
+		fallthrough
+	case "prometheus":
+		endpoint := "/metrics"
+		if cfg.Metrics.Prometheus.Endpoint != nil {
+			endpoint = *cfg.Metrics.Prometheus.Endpoint
+		}
+		log.Info().
+			Str("endpoint", endpoint).
+			Msg("configuring prometheus")
+		promlog := log.With().
+			Str("component", "metrics-exporter").
+			Logger()
+		pipeline, hf, err := prometheus.InstallNewPipeline(prometheus.Config{
+			OnError: func(err error) {
+				promlog.Error().Err(err).Msg("prometheus error")
+			},
+		})
+		if err != nil {
+			return nil, err
+		}
+		srv.RegisterOnShutdown(pipeline.Stop)
+		mux.HandleFunc(endpoint, hf)
+	case "dogstatsd":
+		log.Info().
+			Str("endpoint", cfg.Metrics.Dogstatsd.URL).
+			Msg("configuring dogstatsd")
+		pipeline, err := dogstatsd.InstallNewPipeline(dogstatsd.Config{
+			URL: cfg.Metrics.Dogstatsd.URL,
+		})
+		if err != nil {
+			return nil, err
+		}
+		srv.RegisterOnShutdown(pipeline.Stop)
+	default:
+	}
+
+	// Trace config
+	traceCfg := sdktrace.Config{
+		// By default, assume an ingress is tagging incoming requests for
+		// tracing.
+		DefaultSampler: sdktrace.NeverSample(),
+	}
+	if cfg.Mode == config.DevMode {
+		// Unless we're in dev mode, then go ham.
+		traceCfg.DefaultSampler = sdktrace.AlwaysSample()
+	}
+	if p := cfg.Trace.Probability; p != nil {
+		traceCfg.DefaultSampler = sdktrace.ProbabilitySampler(*p)
+	}
+	traceOpts := []sdktrace.ProviderOption{
+		sdktrace.WithConfig(traceCfg),
+	}
+	if n := cfg.Trace.Name; n != "" {
+		log.Info().Str("sink", n).Msg("configuring trace sink")
+	}
+Trace:
+	switch cfg.Trace.Name {
+	case "stdout":
+		exporter, err := stdout.NewExporter(stdout.Options{})
+		if err != nil {
+			return nil, err
+		}
+		traceOpts = append(traceOpts, sdktrace.WithSyncer(exporter))
+	case "jaeger":
+		jcfg := &cfg.Trace.Jaeger
+		var e jaeger.EndpointOption
+		lev := log.Info()
+		switch {
+		case jcfg.Agent.Endpoint != "":
+			lev = lev.Str("agent", jcfg.Agent.Endpoint)
+			e = jaeger.WithAgentEndpoint(jcfg.Agent.Endpoint)
+		case jcfg.Collector.Endpoint != "":
+			var opt []jaeger.CollectorEndpointOption
+			u, p := jcfg.Collector.Username, jcfg.Collector.Password
+			if u != nil {
+				lev = lev.Str("username", *u)
+				opt = append(opt, jaeger.WithUsername(*u))
+			}
+			if p != nil {
+				lev = lev.Str("password", strings.Repeat("*", len(*p)))
+				opt = append(opt, jaeger.WithPassword(*p))
+			}
+			e = jaeger.WithCollectorEndpoint(jcfg.Collector.Endpoint, opt...)
+		default:
+			lev.Msg("neither jaeger collector nor agent specified")
```

Comment:
Typically, Jaeger Clients default to `localhost:6831` (UDP) when no explicit option is set. You probably want something like this (although, you'll probably want to store the default value somewhere else):

```go
switch {
case jcfg.Collector.Endpoint != "":
	var opt []jaeger.CollectorEndpointOption
	u, p := jcfg.Collector.Username, jcfg.Collector.Password
	if u != nil {
		lev = lev.Str("username", *u)
		opt = append(opt, jaeger.WithUsername(*u))
	}
	if p != nil {
		lev = lev.Str("password", strings.Repeat("*", len(*p)))
		opt = append(opt, jaeger.WithPassword(*p))
	}
	e = jaeger.WithCollectorEndpoint(jcfg.Collector.Endpoint, opt...)
default:
	endpoint := jcfg.Agent.Endpoint
	if len(endpoint) == 0 {
		endpoint = "localhost:6831"
	}
	lev = lev.Str("agent", jcfg.Agent.Endpoint)
	e = jaeger.WithAgentEndpoint(jcfg.Agent.Endpoint)
}
```